exporter/splunkhecexporter/tracedata_to_splunk.go

```diff
@@ -24,34 +24,42 @@ import (
 )
 
 func traceDataToSplunk(logger *zap.Logger, data pdata.Traces, config *Config) ([]*splunk.Event, int) {
-	octds := internaldata.TraceDataToOC(data)
+
 	numDroppedSpans := 0
 	splunkEvents := make([]*splunk.Event, 0, data.SpanCount())
-	for _, octd := range octds {
-		var host string
-		if octd.Resource != nil {
-			host = octd.Resource.Labels[conventions.AttributeHostHostname]
-		}
-		if host == "" {
-			host = unknownHostName
-		}
-		for _, span := range octd.Spans {
-			if span.StartTime == nil {
-				logger.Debug(
-					"Span dropped as it had no start timestamp",
-					zap.Any("span", span))
-				numDroppedSpans++
-				continue
+	for i := 0; i < data.ResourceSpans().Len(); i++ {
+		rs := data.ResourceSpans().At(i)
+		host := unknownHostName
+		source := config.Source
+		sourceType := config.SourceType
+		if !rs.Resource().IsNil() {
+			if conventionHost, isSet := rs.Resource().Attributes().Get(conventions.AttributeHostHostname); isSet {
+				host = conventionHost.StringVal()
+			}
+			if sourceSet, isSet := rs.Resource().Attributes().Get(conventions.AttributeServiceName); isSet {
+				source = sourceSet.StringVal()
 			}
-			se := &splunk.Event{
-				Time:       timestampToEpochMilliseconds(span.StartTime),
-				Host:       host,
-				Source:     config.Source,
-				SourceType: config.SourceType,
-				Index:      config.Index,
-				Event:      span,
+			if sourcetypeSet, isSet := rs.Resource().Attributes().Get(splunk.SourcetypeLabel); isSet {
+				sourceType = sourcetypeSet.StringVal()
+			}
+		}
+		for sils := 0; sils < rs.InstrumentationLibrarySpans().Len(); sils++ {
+			ils := rs.InstrumentationLibrarySpans().At(sils)
+			for si := 0; si < ils.Spans().Len(); si++ {
+				span := ils.Spans().At(si)
+				// TODO replace with span JSON object.
+				spanOC := internaldata.TraceDataToOC(data)[i].Spans[(sils+1)*si]
```

Comment:
TraceDataToOC is an expensive function. Calling it repeatedly just to get a small portion of the data is inefficient. This likely makes things worse than it was before performance wise. I suggest that we either explicitly convert the data to JSON or wait until interndata exposes the necessary function (unclear yet what exactly we need though that can be expressed as a generic functionality).

Reply:
Understood - sounds like at the very least I can move the conversion outside the for loop, and have done so.
This changeset can wait for a better way to express spans as JSON, indeed.
It's just easier for a slew of changes I have queued up - I'll document.

Reply:
If we keep the `TraceDataToOC` call anywhere (even a single call outside the loop) then we are not really gaining much here, do we? The dependency on the `internaldata` is still here and likely performance-wise it is going to be overall similar. So I am not sure what the benefit of the change really. I may be missing something.

Reply:
@atoulme any thoughts on this? Should we table this for now?

Reply:
We can table this for now.
I wanted it out of the way so we could start working on other aspects, such as passthrough of index attribute. It would also help with e2e testing so we could check we get the same pdata and splunk events between transformations.